src/COVID19/model.py

```diff
@@ -345,8 +345,20 @@ def _create(self):
         """
         LOGGER.info("Started model creation")
         self.c_model = covid19.create_model(self.c_params)
-        LOGGER.info("Successfuly created model")
-
+        LOGGER.info("Successfuly created model")    
+    
+    def _destroy(self):
```

Comment:
destroy model is part of the swig interface

you'll see in `model_utils.i`

```
%extend model{
    ~model() {
        destroy_model($self);
    }
}

```
this ensures that 
in the autogenerated code  `__swig_destroy__ = _covid19.delete_model` gets called

which cascades to 

```
SWIGINTERN void delete_model(struct model *self){
        destroy_model(self);
    }
```

If you call the destroy model manually you'll get a seg fault during the double free as the model object goes out of scope.

Reply:
ahh ok, I think I'm a bit confused by this, how do I call this `destroy_model` function from the swig interface? Or are you saying that this doesn't need to be manually called from python?

Interestingly I didn't seem to be getting a segfault when destroying the mode manually, I think it's because the way the `Simulation` class was set up it means that the model object never goes out of scope? Again not completely sure about this.

Reply:
It shouldn't need to be called from the python interface at all, it should be added to the `__del__` method internally in the python.

Reply:
I spent a long time confused about this and finding out how to extend the method to add the proper destructor was a pain in the ass!

Reply:
To test there's a multi threaded run in example, check if your branch works with this

Reply:
OK this makes more sense that it should go in the `__del__` python destructor, but I'm still unsure which python class should have this destructor. 

I think there's still an issue in that the memory for the model is allocated by calling `self.model._create()` inside `Environment.start_simulation()`, and is stored in `self.c_model` inside the python `Model` class. Both the `Environment` and the `Model` class are passed into `Simulation` and are kept around for the duration of the `Simulation` class, even though you can create new models (and therefore allocate new memory) within the `Simulation`. So I don't think there is a python class that we can give a destructor that will be called while the `Simulation` class is still instantiated. Unless I'm missing something here?

Or an alternative would be to make it so that the `Simulation` class can only ever run one simulation, eg. only create one `c_model`, then presumably everything would be freed when this Simulation class is torn down.

Reply:
incidentally I just tried `multi_run_simulator.py` (which is really neat btw), and it does seem to still be leaking memory, so we clearly need a proper destructor. However I tried to add a `__del__` method to the `Model` class which calls `self._destroy()`, however it caused a segfault as soon as the first thread finished running its timesteps

Reply:
Yeah I found adding to the `__del__` caused a seg fault. Maybe there's a bit of C protection we can add, so we check if the memory is assigned in the destroy function before freeing it.

Reply:
OK I've done some more investigation. It seems to be an issue specific to multi-threading, when the model run ends it's calling the destructor which frees the memory, then because of how the thread pool manages memory it seems to try to and access the memory after it has been freed, possibly in the next thread it starts. It doesn't seem to be calling the `__del__` and therefore `destroy()` twice, so I'm not sure if checking in the c code would work.

However I did discover that if you use _multiprocessing_ instead of multithreading is doesn't seem to cause a problem, eg.

```
from concurrent.futures import ProcessPoolExecutor
```

then replace the Threadpool code with:
```
    with ProcessPoolExecutor(n_threads) as ex:
        outputs = list(
            progress_monitor(
                ex.map(run_model, parameter_set_list)
                , total=len(parameter_set_list)
                , desc="Batch progress"
            )
        )
```

With multiprocessing it calls the destructor correctly and there are no memory leaks. The performance seems equally good (I haven't done thorough testing though), however the progress bars don't seem to show the progress for each process at the same time.

If we are happy switching to multi-processing instead of multi-threading (which is more memory safe in general) then I can add this change to the PR?

Reply:
Yes please do that!
Thanks for doing the investigation Rory!

Reply:
My pleasure! Happy to help out however I can.

I've just pushed a commit with has the multi-process instead of multi-threading, and also adds a `__del__` method to the model.py class.

I've kept the `end_simulation()` method I added to the simulation class, since it is still required, as you can create a new model in there without losing the reference to the old one, which still leaks memory.

Let me know if there are any other suggestions. Otherwise feel free to merge if you're happy with this